src/dictarray.c

```diff
@@ -0,0 +1,536 @@
+        if (metadata->rehashing_node)
+            metadata->rehashing_node = NULL;
+        dictRelease(daGetDict(da, didx));
+    }
+    zfree(da->dicts);
+
+    listRelease(da->state.rehashing);
+    if (da->state.dict_size_index)
+        zfree(da->state.dict_size_index);
+
+    zfree(da);
+}
+
+unsigned long long int daSize(dictarray *da) {
+    if (da->num_dicts != 1) {
+        return da->state.key_count;
+    } else {
+        return dictSize(da->dicts[0]);
+    }
+}
+
+/* This method provides the cumulative sum of all the dictionary buckets
+ * across dictionaries in a database. */
+unsigned long daBuckets(dictarray *da) {
+    if (da->num_dicts != 1) {
+        return da->state.bucket_count;
+    } else {
+        return dictBuckets(da->dicts[0]);
+    }
+}
+
+size_t daMemUsage(dictarray *da) {
+    size_t mem = 0;
+    unsigned long long keys_count = daSize(da);
+    mem += keys_count * dictEntryMemUsage() +
+           daBuckets(da) * sizeof(dictEntry*) +
+           da->num_dicts * (sizeof(dict) + dictMetadataSize(daGetDict(da, 0)));
+    return mem;
+}
+
+/*
+ * This method is used to iterate over the elements of the entire dictarray specifically across dicts.
+ * It's a three pronged approach.
+ *
+ * 1. It uses the provided cursor `cursor` to retrieve the dict index from it.
+ * 2. If the dictionary is in a valid state checked through the provided callback `dictScanValidFunction`,
+ *    it performs a dictScan over the appropriate `keyType` dictionary of `db`.
+ * 3. If the dict is entirely scanned i.e. the cursor has reached 0, the next non empty dict is discovered.
+ *    The dict information is embedded into the cursor and returned.
+ *
+ * To restrict the scan to a single cluster dict, pass a valid dict index as
+ * 'onlydidx', otherwise pass -1.
+ */
+unsigned long long daScan(dictarray *da, unsigned long long cursor,
+                          int onlydidx, dictScanFunction *scan_cb,
+                          dictarrayScanShouldSkipDict *skip_cb,
+                          void *privdata)
+{
+    unsigned long long _cursor = 0;
+    /* During main dictionary traversal in cluster mode, 48 upper bits in the cursor are used for positioning in the HT.
+     * Following lower bits are used for the dict index number, ranging from 0 to 2^num_dicts_bits-1.
+     * Dict index is always 0 at the start of iteration and can be incremented only if there are multiple dicts. */
+    int didx = getAndClearDictIndexFromCursor(da, &cursor);
+    if (onlydidx >= 0) {
+        if (didx < onlydidx) {
+            /* Fast-forward to onlydidx. */
+            assert(onlydidx < da->num_dicts);
+            didx = onlydidx;
+            cursor = 0;
+        } else if (didx > onlydidx) {
+            /* The cursor is already past onlydidx. */
+            return 0;
+        }
+    }
+
+    dict *d = daGetDict(da, didx);
+
+    int skip = skip_cb && skip_cb(d);
+    if (!skip) {
+        _cursor = dictScan(d, cursor, scan_cb, privdata);
```

Comment:
we need to make sure that `_cursor` doesn't use more than `64-num_dicts_bits` bits, otherwise `addDictIndexToCursor` will mess up

Reply:
i think we should add some code to detect this case, and then fail SCAN with some error.
since it's a runtime error, and not a logical error in the code, i don't think killing the process with an assert is right.
@madolson WDYT?

Reply:
for the record, this problem exists in unstable today

Reply:
Yeah, we are relying on some assumption that we can only support 2^48 items in a dictionary, which is far beyond what we could practically support. (The dictionary would be at least 1PB in side alone) We could consider adding a static assert here to bound the size of the number of dicts, just so it's clearer.

Reply:
makes sense (to add a an assert on the number of dicts).. i didn't bother to do the math.